ships/0029-one-off-builds.md

```diff
@@ -0,0 +1,301 @@
+
+When an embedded specification is invalid, we propose not to start a `TaskRun`. Instead, the buildrun should be set
+to a failure state and the reasons for the invalid specification should be reflected using the `Succeded` condition in `status.conditions`.
+
+Additionally, we require that `buildSpec` is never used in conjunction with any of `buildRef`, `output`, `paramValues`,
+`env`, `timeout` or `sources`.
+When there is `buildRef` and `buildSpec`, one might decide to prefer `buildSpec` due to locality
+properties. However, it would leave a dead property in the resource description, hurting its declarative nature and
+adding the additional overhead of implementation details for the user to remember. So to
+prevent confusion in the first place, a validation functions should fail the creation of such buildrun and inform the
+user about the exclusive usage of those properties.
+Properties like `output` and `sources`, that overwrite the referenced build's properties, have no usage when an embedded
+build specification is set. Instead, these properties can be directly specified inside `buildSpec`.
+To keep the user from guessing implementation details, a validation function
+should fail the creation of such buildrun again and advise the user to directly set the fields within the embedded
+build specification.
+
+So the following examples are invalid resource definitions:
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata:
+    generateName: buildpack-nodejs-buildrun-
+spec:
+    buildRef:
+        name: buildpack-nodejs-build
+    buildSpec:
+        ...
+```
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata:
+    generateName: buildpack-nodejs-buildrun-
+spec:
+    output:
+        image: my-user/nodejs-build
+    buildSpec:
+        ...
+```
+
+### User Stories
+
+#### Story 1
+
+As a shipwright user, I want to be able to build a container image without creating and managing multiple resources.
+
+#### Story 2
+
+As an application developer, I want my application to utilize `shipwright/build` without the need to create builds only for
+the sake of creating buildruns since the re-usability of builds does not provide any advantages for my application.
+
+#### Story 3
+
+As a developer that uses shipwright, I want to build an image from my local source-code ergonomically in a single step without having
+to maintain multiple resources similar to `docker build`.
+
+### Implementation Notes
+
+#### API Changes
+
+The proposal adds the new optional field `buildSpec` to the `BuildRun` resource specification. It also changes the
+`buildRef` field to optional, since either a `buildSpec` or a `buildRef` would be required to run a build:
+
+```go
+type BuildRunSpec struct {
+    // BuildSpec refers to an embedded build specification
+    // + optional
+    BuildSpec BuildSpec `json:"buildSpec,omitempty"`
+
+    // BuildRef refers to the Build.
+    // +optional
+    BuildRef BuildRef `json:"buildRef,omitempty"`
```

Comment:
Side note - technically this is not a breaking change. Moving a field from required to optional will not impact existing `BuildRun` objects.